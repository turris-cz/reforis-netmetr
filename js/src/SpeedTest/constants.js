/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

const IS_LOCALE_CS = ForisTranslations.locale === "cs";

const NETMETR_BASE_URL = IS_LOCALE_CS ? "https://www.netmetr.cz/cs/" : "https://www.netmetr.cz/en/";

const NETMETR_MY_URL = IS_LOCALE_CS ? `${NETMETR_BASE_URL}moje.html` : `${NETMETR_BASE_URL}my.html`;
const NETMETR_DETAIL_URL = `${NETMETR_BASE_URL}detail.html`;

export { NETMETR_MY_URL, NETMETR_DETAIL_URL };
