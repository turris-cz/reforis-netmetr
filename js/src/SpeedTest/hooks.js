/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useEffect } from "react";
import { ALERT_TYPES, useAlert, useWSForisModule } from "foris";

export default function useNetmetrAlerts(ws) {
    const [setAlert] = useAlert();

    const [downloadDataFinishedData] = useWSForisModule(ws, "netmetr", "download_data_finished");
    useEffect(() => {
        if (downloadDataFinishedData) {
            if (downloadDataFinishedData.passed) {
                setAlert(_("Data downloaded successfully."), ALERT_TYPES.SUCCESS);
            } else {
                setAlert(_("Data downloading failed."), ALERT_TYPES.DANGER);
            }
        }
    }, [downloadDataFinishedData, setAlert]);

    const [measureAndDownloadDataFinishedData] = useWSForisModule(ws, "netmetr", "measure_and_download_data_finished");
    useEffect(() => {
        if (measureAndDownloadDataFinishedData) {
            if (measureAndDownloadDataFinishedData.passed) {
                setAlert(_("Speed test finished successfully."), ALERT_TYPES.SUCCESS);
            } else {
                setAlert(_("Speed test failed."), ALERT_TYPES.DANGER);
            }
        }
    }, [measureAndDownloadDataFinishedData, setAlert]);
}
