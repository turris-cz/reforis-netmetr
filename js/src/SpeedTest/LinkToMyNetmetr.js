/*
 * Copyright (C) 2020-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { NETMETR_MY_URL } from "./constants";

LinkToMyNetmetr.propTypes = {
    syncCode: PropTypes.string.isRequired,
};

export default function LinkToMyNetmetr({ syncCode }) {
    return (
        <p
            dangerouslySetInnerHTML={{
                __html: _(
                    `For more information you need to enter your sync code <strong>${syncCode}</strong> <a href="${NETMETR_MY_URL}" target="_blank" rel="noopener noreferer">here<sup><i class="fas fa-external-link-alt fa-xs ml-1"></i></sup></a>.`
                ),
            }}
        />
    );
}
