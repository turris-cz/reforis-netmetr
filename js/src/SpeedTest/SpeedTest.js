/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState, useEffect } from "react";
import { useAPIGet } from "foris";
import PropTypes from "prop-types";

import TestProgress from "./TestProgress/TestProgress";
import Results from "./Results/Results";
import LinkToMyNetmetr from "./LinkToMyNetmetr";
import NetMetr from "../NetMetr";
import useNetmetrAlerts from "./hooks";
import API_URLs from "../API";
import StartTestButton from "./ControlButtons/StartTestButton";

SpeedTest.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function SpeedTest({ ws }) {
    useNetmetrAlerts(ws);
    const [asyncIdSpeedTest, setAsyncIdSpeedTest] = useState(null);
    const [asyncIdRedownloadData, setAsyncIdRedownloadData] = useState(null);

    const [getSettingsState, getSettingRequest] = useAPIGet(API_URLs.settings);
    useEffect(() => {
        getSettingRequest();
    }, [getSettingRequest]);

    return (
        <NetMetr>
            <div className="card p-4">
                <h2>{_("Speed Test")}</h2>
                <div className="row mb-2">
                    <div className="col-12 col-md-4 col-lg-2">
                        <StartTestButton
                            ws={ws}
                            asyncId={asyncIdSpeedTest}
                            setAsyncId={setAsyncIdSpeedTest}
                        />
                    </div>
                    <div className="col-12 col-md-8 col-lg-10">
                        <TestProgress ws={ws} asyncId={asyncIdSpeedTest} />
                    </div>
                </div>
                <Results
                    ws={ws}
                    asyncIdRedownloadData={asyncIdRedownloadData}
                    setAsyncIdRedownloadData={setAsyncIdRedownloadData}
                />
                {getSettingsState.data && (
                    <LinkToMyNetmetr
                        syncCode={getSettingsState.data.sync_code}
                    />
                )}
            </div>
        </NetMetr>
    );
}
