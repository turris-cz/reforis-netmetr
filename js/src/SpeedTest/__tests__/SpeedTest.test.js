/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { act, render, wait } from "foris/testUtils/customTestRender";
import { ALERT_TYPES, WebSockets } from "foris";
import { mockSetAlert } from "foris/testUtils/alertContextMock";

import diffSnapshot from "snapshot-diff";
import mockAxios from "jest-mock-axios";

import SpeedTest from "../SpeedTest";
import resultsFixture from "../Results/__tests__/__fixtures__/results";

describe("<SpeedTest />", () => {
    const webSockets = new WebSockets();
    let container;
    let asFragment;
    let getByText;
    let firstRender;

    beforeEach(async () => {
        ({ container, asFragment, getByText } = render(
            <SpeedTest ws={webSockets} />
        ));
        mockAxios.mockResponse({ data: resultsFixture });
        await wait(() => getByText("Speed Test"));
        firstRender = asFragment();
    });

    it("Should render component.", () => {
        expect(firstRender).toMatchSnapshot();
    });

    it("Should show success alert when test is finished successfully.", async () => {
        act(() =>
            webSockets.dispatch({
                module: "netmetr",
                action: "measure_and_download_data_finished",
                data: {
                    async_id: "other ID",
                    passed: true,
                },
            })
        );
        await wait(() => {
            expect(mockSetAlert).toBeCalledWith(
                "Speed test finished successfully.",
                ALERT_TYPES.SUCCESS
            );
        });
    });

    it("Should show danger alert when test is finished unsuccessfully.", async () => {
        act(() =>
            webSockets.dispatch({
                module: "netmetr",
                action: "measure_and_download_data_finished",
                data: {
                    async_id: "other ID",
                    passed: false,
                },
            })
        );
        await wait(() => {
            expect(mockSetAlert).toBeCalledWith(
                "Speed test failed.",
                ALERT_TYPES.DANGER
            );
        });
    });
});
