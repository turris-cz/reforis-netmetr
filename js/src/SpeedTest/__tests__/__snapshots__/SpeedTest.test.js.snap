// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpeedTest /> Should render component. 1`] = `
<DocumentFragment>
  <h1>
    NetMetr
  </h1>
  <p>
    NetMetr measures your internet parameters like upload, download and response time.
  </p>
  <div
    class="card p-4"
  >
    <h2>
      Speed Test
    </h2>
    <div
      class="row mb-2"
    >
      <div
        class="col-12 col-md-4 col-lg-2"
      >
        <button
          class="btn btn btn-primary col-sm-12 mb-2"
          id="start-test-button"
          type="button"
        >
          Start test
        </button>
      </div>
      <div
        class="col-12 col-md-8 col-lg-10"
      >
        <div
          class="progress mb-2"
        >
          <div
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="0"
            class="progress-bar progress-bar-striped progress-bar-animated bg-success text-uppercase"
            id="progressbar"
            role="progressbar"
            style="width: 0%; font-size: 11px;"
          />
        </div>
      </div>
    </div>
    <h2>
      Results
      <button
        class="btn btn btn-outline-primary float-right"
        id="redownload-data-button"
        type="button"
      >
        Redownload data
      </button>
    </h2>
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover text-center"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th
              scope="col"
            >
              Date and Time
            </th>
            <th
              scope="col"
            >
              Download [Mb/s]
            </th>
            <th
              scope="col"
            >
              Upload [Mb/s]
            </th>
            <th
              scope="col"
            >
              Ping [ms]
            </th>
            <th
              scope="col"
            >
              Link
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              4/8/2020 5:09 PM
            </td>
            <td>
              30
            </td>
            <td>
              29
            </td>
            <td>
              0.89
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-1"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 4:11 PM
            </td>
            <td>
              30
            </td>
            <td>
              29
            </td>
            <td>
              0.81
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-2"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 3:17 PM
            </td>
            <td>
              30
            </td>
            <td>
              29
            </td>
            <td>
              0.81
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-3"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 2:30 PM
            </td>
            <td>
              30
            </td>
            <td>
              29
            </td>
            <td>
              0.81
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-4"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 1:01 PM
            </td>
            <td>
              30
            </td>
            <td>
              30
            </td>
            <td>
              0.86
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-5"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 12:16 PM
            </td>
            <td>
              30
            </td>
            <td>
              29
            </td>
            <td>
              0.83
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-6"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 11:01 AM
            </td>
            <td>
              30
            </td>
            <td>
              29
            </td>
            <td>
              0.84
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-7"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 10:03 AM
            </td>
            <td>
              30
            </td>
            <td>
              29
            </td>
            <td>
              0.84
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-8"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 9:18 AM
            </td>
            <td>
              30
            </td>
            <td>
              30
            </td>
            <td>
              0.84
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-9"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
          <tr>
            <td>
              4/8/2020 8:05 AM
            </td>
            <td>
              30
            </td>
            <td>
              29
            </td>
            <td>
              0.83
            </td>
            <td>
              <a
                href="https://www.netmetr.cz/en/detail.html?1234-1234-1234-1234-10"
                rel="noopener noreferrer"
                target="_blank"
              >
                Details
                <sup>
                  <i
                    class="fas fa-external-link-alt ml-1"
                  />
                </sup>
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
